ComponentKit/Core/CKComponent.mm

```diff
@@ -275,21 +275,11 @@ + (id)initialState
   return nil;
 }
 
-- (void)updateState:(id (^)(id))updateBlock
-{
-  [self _updateState:updateBlock tryAsynchronousUpdate:NO];
-}
-
-- (void)updateStateWithExpensiveReflow:(id (^)(id))updateBlock
-{
-  [self _updateState:updateBlock tryAsynchronousUpdate:YES];
-}
-
-- (void)_updateState:(id (^)(id))updateBlock tryAsynchronousUpdate:(BOOL)tryAsynchronousUpdate
+- (void)updateState:(id (^)(id))updateBlock mode:(CKUpdateMode)mode
```

Comment:
Should we have a stateUpdate method with default behavior (ie everything is async by default) ?

== Discussion around the API within components ==

Specifying sync and async from the component is definitely the most straightforward approach but it makes me a bit sad because:
- It couples components to a certain implementation of an hosting view and data source.
- It can be easy to take strange assumptions around the fact that we guarantee some state updates will be sync (already the case today though).
- It will prevent enforcing different policies than sync and async (e.g we could do our best for a state update to be sync but still scatter it across different run loops if needed). This point is far down the road I agree though. 

It feels like the decision between sync and async lies around two semantic factors :
- Prioritity of the interaction (should be defined by the component)
- Complexity of the state update (should be inferred by the infra)

What if we exposed in components land an API based on priority ? Inferring complexity is tricky (number of components in the tree) and depends on the context the component is presented in (e.g button in an h-scroll vs same button in a simple story) so we could assume homogeneous complexity for now and just reason in terms of priority of the interaction (e.g button tap high pri, state update due to incoming data coming from the server low-pri).

Thoughts ?